src/main/java/org/elasticsearch/search/lookup/SourceLookup.java

```diff
@@ -99,34 +99,26 @@ public XContentType sourceContentType() {
         return sourceAsMapAndType(bytes, offset, length).v2();
     }
 
-    public void setNextReader(LeafReaderContext context) {
-        if (this.reader == context.reader()) { // if we are called with the same reader, don't invalidate source
+    public void setSource(LeafReaderContext context, int docId) {
+        if (this.reader == context.reader() && this.docId == docId) {
+            // if we are called with the same docId, don't invalidate source
```

Comment:
maybe say document instead of docId? since the check is technically same docId in the same segment?